File: test/python/helper/merchant_bni_va_topup.py
import json
import os
import threading
from datetime import datetime, timedelta, timezone

import urllib.request

from dana.api_client import ApiClient
from dana.merchant_management.v1 import MerchantManagementApi
from dana.merchant_management.v1.models.query_asset_card_list_request import QueryAssetCardListRequest
from dana.merchant_management.v1.models.query_merchant_info_request import QueryMerchantInfoRequest
from dana.utils.snap_configuration import AuthSettings, Env, SnapConfiguration
from helper.bni_hash_util import hash_data

DANA_SANDBOX_BASE_URL = "https://api.sandbox.dana.id"
BNI_VA_TOP_UP_PATH = "/ifcsupergw/bni/topup/merchant/request.htm"

DEFAULT_BNI_CLIENT_ID = "910"
DEFAULT_BNI_SECRET_KEY = "9546d5f69af2ed3bc603834446985628"
BNI_INST_ID = "BNIC1ID"
MERCHANT_DEPOSIT_ACCOUNT_TYPE = "MERCHANT_DEPOSIT_ACCOUNT"
MERCHANT_QUERY_LOGIN_TYPE_ROLE = "ROLE"
MERCHANT_DEPOSIT_TOP_UP_THRESHOLD = 1_000_000

_lock = threading.Lock()
_done = False
_failure: Exception | None = None
_merchant_api: MerchantManagementApi | None = None


def ensure_merchant_bni_va_topup() -> None:
    global _done, _failure
    if _done:
        if _failure is not None:
            raise _failure
        return
    with _lock:
        if _done:
            if _failure is not None:
                raise _failure
            return
        try:
            _bni_va_topup_merchant_once()
        except Exception as exc:
            _failure = exc
            raise
        finally:
            _done = True


def _get_merchant_management_api() -> MerchantManagementApi:
    global _merchant_api
    if _merchant_api is None:
        configuration = SnapConfiguration(
            api_key=AuthSettings(
                PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
                ORIGIN=os.environ.get("ORIGIN"),
                X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
                DANA_ENV=Env.SANDBOX,
                X_DEBUG="true",
                CLIENT_SECRET=os.environ.get("CLIENT_SECRET"),
            )
        )
        _merchant_api = MerchantManagementApi(ApiClient(configuration))
    return _merchant_api


def _bni_va_topup_merchant_once() -> None:
    deposit_balance = _query_merchant_deposit_total_amount()
    if deposit_balance >= MERCHANT_DEPOSIT_TOP_UP_THRESHOLD:
        print(
            f"Skipping BNI VA top-up: merchant deposit balance={deposit_balance} "
            f">= threshold={MERCHANT_DEPOSIT_TOP_UP_THRESHOLD}"
        )
        return
    print(
        f"Merchant deposit balance={deposit_balance} < threshold={MERCHANT_DEPOSIT_TOP_UP_THRESHOLD}; "
        "proceeding with BNI VA top-up"
    )

    virtual_account = _query_bni_merchant_virtual_account()
    _post_bni_va_topup_merchant(virtual_account)


def _query_merchant_deposit_total_amount() -> int:
    merchant_id = os.getenv("MERCHANT_ID", "")
    if not merchant_id:
        raise RuntimeError("MERCHANT_ID is required to query merchant info")

    request = QueryMerchantInfoRequest(
        role_id=merchant_id,
        login_type=MERCHANT_QUERY_LOGIN_TYPE_ROLE,
        is_query_account=True,
    )
    response = _get_merchant_management_api().query_merchant_info(request)

    result_info = response.response.body.result_info
    if result_info.result_status != "S":
        raise RuntimeError(f"queryMerchantInfo failed: {result_info}")

    accounts = getattr(response.response.body.merchant_information, "accounts", None) or []
    for account in accounts:
        if account.account_type == MERCHANT_DEPOSIT_ACCOUNT_TYPE:
            return _parse_account_mapped_total_amount(account)

    raise RuntimeError(
        f"queryMerchantInfo: {MERCHANT_DEPOSIT_ACCOUNT_TYPE} account not found"
    )


def _parse_account_mapped_total_amount(account) -> int:
    mapped_total = getattr(account, "mapped_total_amount", None)
    if mapped_total is not None and getattr(mapped_total, "amount", None) is not None:
        return _parse_amount_value(mapped_total.amount)

    account_dict = account.to_dict() if hasattr(account, "to_dict") else {}
    mapped_total = account_dict.get("mappedTotalAmount")
    if isinstance(mapped_total, dict) and "amount" in mapped_total:
        return _parse_amount_value(mapped_total["amount"])

    total_amount = getattr(account, "total_amount", None) or account_dict.get("totalAmount", "")
    if isinstance(total_amount, str) and total_amount:
        parsed = json.loads(total_amount)
        return _parse_amount_value(parsed.get("amount"))

    raise RuntimeError("deposit account amount missing in queryMerchantInfo response")


def _parse_amount_value(value) -> int:
    if isinstance(value, str):
        return int(value)
    if isinstance(value, bool):
        raise RuntimeError(f"unsupported amount format: {value!r}")
    if isinstance(value, (int, float)):
        return int(value)
    raise RuntimeError(f"unsupported amount format: {value!r}")


def _query_bni_merchant_virtual_account() -> str:
    member_id = os.getenv("MERCHANT_ID", "")
    if not member_id:
        raise RuntimeError("MERCHANT_ID is required to query merchant BNI VA")

    request = QueryAssetCardListRequest(
        member_id=member_id,
        enable_only="true",
        asset_type_list=["VA_ACCOUNT"],
    )
    response = _get_merchant_management_api().query_asset_card_list(request)

    result_info = response.response.body.result_info
    if result_info.result_status != "S":
        raise RuntimeError(f"queryAssetCardList failed: {result_info}")

    for card in response.response.body.asset_card_list or []:
        if card.asset_type == "VA_ACCOUNT" and card.inst_id == BNI_INST_ID:
            if card.card_index_no:
                return card.card_index_no

    raise RuntimeError(
        f"BNI VA card not found in assetCardList (assetType=VA_ACCOUNT instId={BNI_INST_ID})"
    )


def _post_bni_va_topup_merchant(virtual_account: str) -> None:
    jakarta = timezone(timedelta(hours=7))
    now = datetime.now(jakarta)
    integration_body = {
        "trx_amount": "1000",
        "trx_id": now.strftime("%Y%m%d%H%M%S"),
        "virtual_account": virtual_account,
        "customer_name": "rudy",
        "payment_amount": "1000000000",
        "cumulative_payment_amount": "1000",
        "payment_ntb": "233171",
        "datetime_payment": now.strftime("%Y-%m-%d %H:%M:%S"),
        "datetime_payment_iso8601": now.strftime("%Y-%m-%dT%H:%M:%S%z"),
    }

    client_id = os.getenv("BNI_VA_TOP_UP_CLIENT_ID", DEFAULT_BNI_CLIENT_ID)
    secret_key = os.getenv("BNI_VA_TOP_UP_SECRET_KEY", DEFAULT_BNI_SECRET_KEY)
    integration_json = json.dumps(integration_body, separators=(",", ":"))
    data = hash_data(integration_json, client_id, secret_key)

    payload = json.dumps({"client_id": client_id, "data": data})
    _http_post_json(DANA_SANDBOX_BASE_URL + BNI_VA_TOP_UP_PATH, payload)


def _http_post_json(url: str, body: str) -> str:
    req = urllib.request.Request(
        url,
        data=body.encode("utf-8"),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=30) as resp:
        response_text = resp.read().decode("utf-8")
        if resp.status < 200 or resp.status >= 300:
            raise RuntimeError(f"HTTP {resp.status}: {response_text}")

    try:
        parsed = json.loads(response_text)
        status = parsed.get("status")
        if status and status != "000":
            raise RuntimeError(f"BNI VA top-up rejected: {response_text}")
    except json.JSONDecodeError:
        pass

    return response_text

File: test/php/helper/MerchantBniVaTopUp.php
<?php

namespace DanaUat\Helper;

use Dana\Configuration;
use Dana\Env;
use Dana\MerchantManagement\v1\Api\MerchantManagementApi;
use Dana\MerchantManagement\v1\Model\QueryAssetCardListRequest;
use Dana\MerchantManagement\v1\Model\QueryMerchantInfoRequest;
use Exception;

final class MerchantBniVaTopUp
{
    private const DANA_SANDBOX_BASE_URL = 'https://api.sandbox.dana.id';
    private const BNI_VA_TOP_UP_PATH = '/ifcsupergw/bni/topup/merchant/request.htm';

    private const DEFAULT_BNI_CLIENT_ID = '910';
    private const DEFAULT_BNI_SECRET_KEY = '9546d5f69af2ed3bc603834446985628';
    private const BNI_INST_ID = 'BNIC1ID';
    private const MERCHANT_DEPOSIT_ACCOUNT_TYPE = 'MERCHANT_DEPOSIT_ACCOUNT';
    private const MERCHANT_QUERY_LOGIN_TYPE_ROLE = 'ROLE';
    private const MERCHANT_DEPOSIT_TOP_UP_THRESHOLD = 1000000;

    private static bool $done = false;
    private static ?Exception $failure = null;
    private static ?MerchantManagementApi $merchantManagementApi = null;

    private function __construct()
    {
    }

    public static function ensure(): void
    {
        if (self::$done) {
            if (self::$failure !== null) {
                throw self::$failure;
            }
            return;
        }

        try {
            self::bniVaTopUpMerchantOnce();
        } catch (Exception $e) {
            self::$failure = $e;
            throw $e;
        } finally {
            self::$done = true;
        }
    }

    private static function bniVaTopUpMerchantOnce(): void
    {
        $depositBalance = self::queryMerchantDepositTotalAmount();
        if ($depositBalance >= self::MERCHANT_DEPOSIT_TOP_UP_THRESHOLD) {
            echo 'Skipping BNI VA top-up: merchant deposit balance=' . $depositBalance
                . ' >= threshold=' . self::MERCHANT_DEPOSIT_TOP_UP_THRESHOLD . PHP_EOL;
            return;
        }
        echo 'Merchant deposit balance=' . $depositBalance
            . ' < threshold=' . self::MERCHANT_DEPOSIT_TOP_UP_THRESHOLD
            . '; proceeding with BNI VA top-up' . PHP_EOL;

        $virtualAccount = self::queryBniMerchantVirtualAccount();
        self::postBniVaTopUpMerchant($virtualAccount);
    }

    private static function queryMerchantDepositTotalAmount(): int
    {
        $merchantId = getenv('MERCHANT_ID') ?: '';
        if ($merchantId === '') {
            throw new Exception('MERCHANT_ID is required to query merchant info');
        }

        $request = new QueryMerchantInfoRequest([
            'roleId' => $merchantId,
            'loginType' => self::MERCHANT_QUERY_LOGIN_TYPE_ROLE,
            'isQueryAccount' => true,
        ]);

        $response = self::getMerchantManagementApi()->queryMerchantInfo($request);
        $resultInfo = $response->getResponse()->getBody()->getResultInfo();
        if ($resultInfo->getResultStatus() !== 'S') {
            throw new Exception('queryMerchantInfo failed: ' . json_encode([
                'resultStatus' => $resultInfo->getResultStatus(),
                'resultCodeId' => $resultInfo->getResultCodeId(),
                'resultMsg' => $resultInfo->getResultMsg(),
            ]));
        }

        $merchantInformation = $response->getResponse()->getBody()->getMerchantInformation();
        if ($merchantInformation === null || $merchantInformation->getAccounts() === null) {
            throw new Exception('queryMerchantInfo: merchantInformation.accounts missing');
        }

        foreach ($merchantInformation->getAccounts() as $account) {
            if ($account->getAccountType() === self::MERCHANT_DEPOSIT_ACCOUNT_TYPE) {
                return self::parseAccountMappedTotalAmount($account);
            }
        }

        throw new Exception('queryMerchantInfo: ' . self::MERCHANT_DEPOSIT_ACCOUNT_TYPE . ' account not found');
    }

    /**
     * @param object $account
     */
    private static function parseAccountMappedTotalAmount($account): int
    {
        $accountData = json_decode(json_encode($account), true);
        if (!is_array($accountData)) {
            throw new Exception('deposit account: invalid account payload');
        }

        $mappedTotalAmount = $accountData['mappedTotalAmount'] ?? null;
        if (is_array($mappedTotalAmount) && array_key_exists('amount', $mappedTotalAmount)) {
            return self::parseAmountValue($mappedTotalAmount['amount']);
        }

        $totalAmount = $accountData['totalAmount'] ?? '';
        if (is_string($totalAmount) && $totalAmount !== '') {
            $parsed = json_decode($totalAmount, true);
            if (is_array($parsed) && array_key_exists('amount', $parsed)) {
                return self::parseAmountValue($parsed['amount']);
            }
        }

        throw new Exception('deposit account amount missing in queryMerchantInfo response');
    }

    private static function parseAmountValue($value): int
    {
        if (is_string($value)) {
            return (int) $value;
        }
        if (is_int($value)) {
            return $value;
        }
        if (is_float($value)) {
            return (int) $value;
        }

        throw new Exception('unsupported amount format: ' . json_encode($value));
    }

    private static function queryBniMerchantVirtualAccount(): string
    {
        $memberId = getenv('MERCHANT_ID') ?: '';
        if ($memberId === '') {
            throw new Exception('MERCHANT_ID is required to query merchant BNI VA');
        }

        $request = new QueryAssetCardListRequest([
            'memberId' => $memberId,
            'enableOnly' => 'true',
            'assetTypeList' => ['VA_ACCOUNT'],
        ]);

        $response = self::getMerchantManagementApi()->queryAssetCardList($request);
        $resultInfo = $response->getResponse()->getBody()->getResultInfo();
        if ($resultInfo->getResultStatus() !== 'S') {
            throw new Exception('queryAssetCardList failed: ' . json_encode([
                'resultStatus' => $resultInfo->getResultStatus(),
                'resultCodeId' => $resultInfo->getResultCodeId(),
                'resultMsg' => $resultInfo->getResultMsg(),
            ]));
        }

        foreach ($response->getResponse()->getBody()->getAssetCardList() ?? [] as $card) {
            if ($card->getAssetType() === 'VA_ACCOUNT' && $card->getInstId() === self::BNI_INST_ID) {
                $cardIndexNo = $card->getCardIndexNo() ?? '';
                if ($cardIndexNo !== '') {
                    return $cardIndexNo;
                }
            }
        }

        throw new Exception('BNI VA card not found in assetCardList');
    }

    private static function postBniVaTopUpMerchant(string $virtualAccount): void
    {
        $now = new \DateTime('now', new \DateTimeZone('Asia/Jakarta'));
        $integrationBody = [
            'trx_amount' => '1000',
            'trx_id' => $now->format('YmdHis'),
            'virtual_account' => $virtualAccount,
            'customer_name' => 'rudy',
            'payment_amount' => '1000000000',
            'cumulative_payment_amount' => '1000',
            'payment_ntb' => '233171',
            'datetime_payment' => $now->format('Y-m-d H:i:s'),
            'datetime_payment_iso8601' => $now->format('Y-m-d\TH:i:sP'),
        ];

        $clientId = getenv('BNI_VA_TOP_UP_CLIENT_ID') ?: self::DEFAULT_BNI_CLIENT_ID;
        $secretKey = getenv('BNI_VA_TOP_UP_SECRET_KEY') ?: self::DEFAULT_BNI_SECRET_KEY;
        $integrationJson = json_encode($integrationBody);
        $data = BniHashUtil::hashData($integrationJson, $clientId, $secretKey);

        $payload = json_encode(['client_id' => $clientId, 'data' => $data]);
        self::httpPostJson(self::DANA_SANDBOX_BASE_URL . self::BNI_VA_TOP_UP_PATH, $payload, true);
    }

    private static function getMerchantManagementApi(): MerchantManagementApi
    {
        if (self::$merchantManagementApi === null) {
            self::$merchantManagementApi = new MerchantManagementApi(null, self::danaConfiguration());
        }

        return self::$merchantManagementApi;
    }

    private static function danaConfiguration(): Configuration
    {
        $configuration = new Configuration();
        $configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        $configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        $configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        $configuration->setApiKey('CLIENT_SECRET', getenv('CLIENT_SECRET'));
        $configuration->setApiKey('ENV', Env::SANDBOX);
        return $configuration;
    }

    private static function httpPostJson(string $url, string $body, bool $checkBniStatus): void
    {
        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => ['Content-Type: application/json'],
            CURLOPT_POSTFIELDS => $body,
            CURLOPT_TIMEOUT => 30,
        ]);

        $responseText = curl_exec($ch);
        $statusCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        if ($responseText === false) {
            $error = curl_error($ch);
            curl_close($ch);
            throw new Exception('HTTP request failed: ' . $error);
        }
        curl_close($ch);

        if ($statusCode < 200 || $statusCode >= 300) {
            throw new Exception('HTTP ' . $statusCode . ': ' . $responseText);
        }

        if ($checkBniStatus) {
            $parsed = json_decode($responseText, true);
            if (is_array($parsed) && isset($parsed['status']) && $parsed['status'] !== '' && $parsed['status'] !== '000') {
                throw new Exception('BNI VA top-up rejected: ' . $responseText);
            }
        }
    }
}
